pkg/wireguard/agent/agent.go

```diff
@@ -88,14 +93,76 @@ func NewAgent(privKeyPath string) (*Agent, error) {
 		peerByNodeName:   map[string]*peerConfig{},
 		nodeNameByNodeIP: map[string]string{},
 		restoredPubKeys:  map[wgtypes.Key]struct{}{},
+		cleanup:          []func(){},
 	}, nil
 }
 
 // Close is called when the agent stops
 func (a *Agent) Close() error {
+	a.RLock()
+	defer a.RUnlock()
+
+	for _, cleanup := range a.cleanup {
+		cleanup()
+	}
+
 	return a.wgClient.Close()
 }
 
+func (a *Agent) initUserspaceDevice(linkMTU int) (netlink.Link, error) {
+	log.WithField(logfields.Hint,
+		"It is highly recommended to use the kernel implementation. "+
+			"See https://www.wireguard.com/install/ for details.").
+		Info("falling back to the WireGuard userspace implementation.")
```

Comment:
nit: Should this be a Warn instead?

Reply:
I wonder if the fallback should be disabled by default. So that when the agent starts up, it hints to the user that a fallback could be used but with mentioned constraints to be aware of?

Reply:
Yes, I considered that too, and I am now also starting to think that we probably should make it opt-in instead of opt-out. It will make the installation slightly more annoying, but it will stop people from running user-mode by accident, which I think can be a footgun due to it's performance and availability downsides.

Reply:
If I understood you correctly, it's that wg would be running as part of the Cilium agent process, so upgrading may shortly interrupt/break connectivity. I think an implicit fallback would not make people aware of it.

Reply:
That's correct, upgrading or configuration changes will interrupt connectivity in user-mode.